sourcetool/pkg/auth/implementation.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+)
+
+//counterfeiter:generate . authenticatorImplementation
+type authenticatorImplementation interface {
+	requestDeviceCode(context.Context) (*DeviceCodeResponse, error)
+	openBrowser(string) error
+	pollForToken(context.Context, string, time.Duration) (string, error)
+	checkTokenStatus(context.Context, string) (string, error)
+	persistToken(token string) error
+	readToken() (string, error)
+}
+
+type defaultImplementation struct{}
+
+// pollForToken polls GitHub periodically until the device flow is done and
+// the token is issued.
+func (di *defaultImplementation) pollForToken(ctx context.Context, deviceCode string, interval time.Duration) (string, error) {
```

Comment:
nit: polling is more of a UI/CLI responsibility. You might not want to call `time.Sleep` here. Consider returning an error with the next interval instead and allow the cmd code to do the polling loop.

Reply:
Good call, I like the proposed division of responsibilities. I'll handle this in a follow-up PR as I don't want to create a huge diff because the latest commit will not merge cleanly on another (much bigger) branch I'm about to check-in on top of this change.